core/assay.js

```diff
@@ -0,0 +1,131 @@
+import harden from '@agoric/harden';
+
+import { insist } from '../util/insist';
+import { mustBeSameStructure, mustBeComparable } from '../util/sameStructure';
+
+// This module treats labels as black boxes. It is not aware
+// of issuers, and so can handle labels whose issuers are merely
+// presences of remote issuers.
+
+// `makeAssayMaker` takes in a 'strategy' object which defines set
+// operations for a particular kind of assay. For instance, for the
+// natAssay, these operations are arithmetic.
+
+// Return an assay, which makes amounts, validates amounts, and
+// provides set operations over amounts. An amount is a pass-by-copy
+// description of some set of erights. An amount has a label and a
+// quantity. All amounts made by the same assay have the same label
+// but differ in quantity.
+//
+// An assay is pass-by-presence, but is not designed to be usefully
+// passed. Rather, we expect each vat that needs to operate on amounts
+// will have its own local assay to do so.
+
+function makeAssayMaker(strategy) {
+  function makeAssay(label) {
+    mustBeComparable(label);
+
+    // The brand represents recognition of the amount as authorized.
+    const brand = new WeakSet();
+
+    const assay = harden({
+      getLabel() {
+        return label;
+      },
+
+      getStrategy() {
+        return strategy;
+      },
+
+      // Given the raw quantity that this kind of amount would label, return
+      // an amount so labeling that quantity.
+      make(allegedQuantity) {
+        const amount = harden({
+          label,
+          quantity: strategy.insistKind(allegedQuantity),
```

Comment:
I don't think this coerces. It doesn't for Nat (which throws an error) and it doesn't for pixelList. Is there somewhere that I've missed where coercion is happening?

Reply:
Since you're code passes the `allegedQuantity` in to this method and uses what this method returns, you are enabling this method to coerce. Although we don't have an example, this was the original design intention.

If we wish this method only to approve or reject, we should not give it the opportunity to coerce. We should write it as

```js
make(allegedQuantity) {
  strategy.insistKind(allegedQuantity);
  const amount = harden({
    label,
    quantity: allegedQuantity,
```

We should either change its name to something that suggests coercion, or we should change the code that calls it so that it isn't able to coerce.

I think having it be able to coerce is better in this case.

Reply:
I wanted to better understand your use of the word 'coerce'. My understanding of 'coerce' is that it is implicit type conversion from typeA to typeB. That doesn't seem to be what is happening here. For instance, if I call `make(3.9)` and `strategy.insistKind` is `Nat`, `make` will error rather than try to make `3.9` into the Nat `3`. By my definition, there is no coercion.

I think you are saying that because `insistKind` returns the `allegedQuantity` if it doesn't error (it is already a Nat), it is doing coercion. Is that right? I don't see how this fits the definition of coercion since no type conversion occurs. 

It also sounds like you might be saying that anything with the name `insist` should not be returning anything. Is that right? If that's the case, would you be ok with calling `insistKind` `insistKindAndReturn` or something similar? It is really doing an insist, but it is also returning a value but only if there is no error.

Reply:
'coerce' here is a bit more general than only type conversion. The point is that since it returns an object, it can modify the object in arbitrary ways. (The assay is the authority on what objects are valid, so the issue isn't that we're vulnerable to it.) Like with the design of the peg, it might substitute an amount with a presence for an issuer for an equivalent amount with a local issuer. 

This is partly about terminology. A function called insistFoo() should return true or false. A function that takes an arg and returns something represents the same thing is more appropriately called coerceFoo().

Reply:
Gotcha. I think that the proper usage of `insistFoo` (following the behavior of the`insist` util) is not that it returns true or false but that it throws or doesn't throw. I just pushed up a change that moves the `insistKind` call up so that it acts more like a guard. The `quantity` is then just set to `allegedQuantity` as Mark had above.